// license-compliance/lib/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isComplianceModeEnabled = exports.getConfiguration = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const cosmiconfig_1 = require("cosmiconfig");
const joi_1 = tslib_1.__importDefault(require("joi"));
const path_1 = tslib_1.__importDefault(require("path"));
const enumerations_1 = require("./enumerations");
const program_1 = require("./program");
const util_1 = require("./util");
const packageName = "license-compliance";
async function getConfiguration(nodeModulesPath) {
    let configExtended = {};
    let configInline = {};
    const explorer = (0, cosmiconfig_1.cosmiconfig)(packageName, { searchStrategy: "global" });
    const configResult = await explorer.search();
    configInline = configResult === null || configResult === void 0 ? void 0 : configResult.config;
    const extendsPath = configInline === null || configInline === void 0 ? void 0 : configInline.extends;
    if (extendsPath) {
        try {
            const c = await explorer.load(path_1.default.join(nodeModulesPath, extendsPath, "index.js"));
            configExtended = (c === null || c === void 0 ? void 0 : c.config) || {};
            delete configInline.extends;
        }
        catch (error) {
            console.error(chalk_1.default.red("Extended configuration error:"), error);
            return null;
        }
    }
    const mergedConfiguration = Object.assign(configExtended, configInline, (0, program_1.processArgs)());
    const configuration = {
        allow: mergedConfiguration.allow || [],
        development: !!mergedConfiguration.development || false,
        direct: mergedConfiguration.direct || false,
        exclude: mergedConfiguration.exclude || [],
        format: (0, util_1.toPascal)(mergedConfiguration.format) || enumerations_1.Formatter.text,
        production: !!mergedConfiguration.production || false,
        query: mergedConfiguration.query || [],
        report: (0, util_1.toPascal)(mergedConfiguration.report) || enumerations_1.Report.summary,
    };
    const result = joi_1.default
        .object({
        allow: joi_1.default.array().items(joi_1.default.string()),
        development: joi_1.default.boolean(),
        direct: joi_1.default.boolean(),
        exclude: joi_1.default.array(),
        format: joi_1.default.string().valid(enumerations_1.Formatter.csv, enumerations_1.Formatter.json, enumerations_1.Formatter.text, enumerations_1.Formatter.xunit),
        production: joi_1.default.boolean(),
        query: joi_1.default.array().items(joi_1.default.string()),
        report: joi_1.default.string().valid(enumerations_1.Report.detailed, enumerations_1.Report.summary),
    })
        .validate(configuration);
    if (result.error) {
        console.error(chalk_1.default.red("Configuration error:"), result.error.message);
        return null;
    }
    return configuration;
}
exports.getConfiguration = getConfiguration;
function isComplianceModeEnabled(configuration) {
    return Array.isArray(configuration.allow) && configuration.allow.length > 0;
}
exports.isComplianceModeEnabled = isComplianceModeEnabled;
