// license-compliance/lib/license.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onlyAllow = exports.isLicenseValid = exports.getLicense = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const debug_1 = tslib_1.__importDefault(require("debug"));
const path_1 = tslib_1.__importDefault(require("path"));
const spdx_expression_parse_1 = tslib_1.__importDefault(require("spdx-expression-parse"));
const spdx_satisfies_1 = tslib_1.__importDefault(require("spdx-satisfies"));
const enumerations_1 = require("./enumerations");
const util = tslib_1.__importStar(require("./util"));
const satisfies = (allowedLicenses, license) => {
    try {
        return (0, spdx_satisfies_1.default)(allowedLicenses, license);
    }
    catch (error) {
        debug(`Failed to parse license identifier "${license}"`, error);
        return false;
    }
};
const debug = (0, debug_1.default)("license-compliance:license");
const SEE_LICENSE_IN = "SEE LICENSE IN";
const LICENSE_FILE = "license";
async function getLicense(pack, packPath) {
    const license = await extractLicense(pack, packPath);
    if (!isLicenseValid(license.name)) {
        license.name = enumerations_1.Literals.UNKNOWN;
        license.status = enumerations_1.LicenseStatus.unknown;
    }
    return license;
}
exports.getLicense = getLicense;
function isLicenseValid(license) {
    if (license === enumerations_1.Literals.CUSTOM) {
        return true;
    }
    if (license === enumerations_1.Literals.UNKNOWN) {
        return false;
    }
    try {
        (0, spdx_expression_parse_1.default)(license);
        return true;
    }
    catch {
        return false;
    }
}
exports.isLicenseValid = isLicenseValid;
function onlyAllow(packages, configuration) {
    if (configuration.allow.length === 0) {
        return [];
    }
    const invalidPackages = new Array();
    const spdxLicense = argsToSpdxLicense(configuration.allow);
    for (const pack of packages) {
        const matches = pack.license !== enumerations_1.Literals.UNKNOWN &&
            pack.license !== enumerations_1.Literals.CUSTOM &&
            satisfies(spdxLicense, pack.license);
        debug(chalk_1.default.blue(pack.name), "/", pack.license, "=>", matches ? chalk_1.default.green(spdxLicense) : chalk_1.default.red(spdxLicense));
        if (!matches) {
            invalidPackages.push(pack);
        }
    }
    return invalidPackages;
}
exports.onlyAllow = onlyAllow;
function argsToSpdxLicense(licenses) {
    if (licenses.length === 1) {
        return licenses[0];
    }
    let buffer = "(";
    for (const license of licenses) {
        buffer += license + " OR ";
    }
    return buffer.substring(0, buffer.length - 4) + ")";
}
async function extractLicense(pack, packPath) {
    const licensePath = await getLicensePath(packPath);
    if (typeof pack.license === "string") {
        if (pack.license.startsWith(SEE_LICENSE_IN)) {
            return {
                name: enumerations_1.Literals.CUSTOM,
                path: await getCustomLicensePath(packPath, pack.license),
                status: enumerations_1.LicenseStatus.custom,
            };
        }
        return {
            name: pack.license,
            path: licensePath,
            status: enumerations_1.LicenseStatus.valid,
        };
    }
    if (pack.license && pack.license.type) {
        return {
            name: pack.license.type,
            path: licensePath,
            status: enumerations_1.LicenseStatus.valid,
        };
    }
    if (Array.isArray(pack.licenses)) {
        return {
            name: getLicenseFromArray(pack.licenses),
            path: null,
            status: enumerations_1.LicenseStatus.valid,
        };
    }
    return {
        name: enumerations_1.Literals.UNKNOWN,
        path: null,
        status: enumerations_1.LicenseStatus.unknown,
    };
}
async function getCustomLicensePath(packPath, license) {
    const licPath = path_1.default.join(packPath, license.substring(SEE_LICENSE_IN.length).trim());
    if (await util.fileExists(licPath)) {
        return licPath;
    }
    return null;
}
function getLicenseFromArray(licenses) {
    if (licenses.length === 0) {
        return enumerations_1.Literals.UNKNOWN;
    }
    if (licenses.length === 1) {
        return licenses[0].type;
    }
    return argsToSpdxLicense(licenses.map((value) => value.type.trim()));
}
async function getLicensePath(packPath) {
    const data = await util.readdir(packPath);
    for (const fileName of data) {
        if (fileName.toLowerCase().startsWith(LICENSE_FILE)) {
            return path_1.default.join(packPath, fileName);
        }
    }
    return null;
}
