// license-compliance/lib/npm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInstalledPackages = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const debug_1 = tslib_1.__importDefault(require("debug"));
const path_1 = tslib_1.__importDefault(require("path"));
const license_1 = require("./license");
const repository_1 = require("./repository");
const util = tslib_1.__importStar(require("./util"));
const debug = (0, debug_1.default)("license-compliance:npm");
const PACKAGE_JSON = "package.json";
const NODE_MODULES = "node_modules";
async function getInstalledPackages(configuration, nodeModulesPath, packageJsonPath = "") {
    const packages = new Array();
    const pack = await util.readPackageJson(path_1.default.join(packageJsonPath, PACKAGE_JSON));
    if (!pack) {
        console.error(chalk_1.default.red(`Could not find '${PACKAGE_JSON}' at '${packageJsonPath}'`));
        return new Array();
    }
    if (pack.dependencies && !configuration.development) {
        debug("Analyzing production dependencies at", nodeModulesPath);
        await readPackages(pack.name, pack.dependencies, 0, nodeModulesPath, configuration, nodeModulesPath, packages);
    }
    if (pack.devDependencies && !configuration.production) {
        debug("Analyzing development dependencies at", nodeModulesPath);
        await readPackages(pack.name, pack.devDependencies, 0, nodeModulesPath, configuration, nodeModulesPath, packages);
    }
    return packages;
}
exports.getInstalledPackages = getInstalledPackages;
function alreadyAnalyzed(packages, pack) {
    return packages.find((value) => value.name === pack.name && value.version === pack.version) !== undefined;
}
async function getInstalledPath(parentName, packageName, parentNodeModulesPath, rootNodeModulesPath) {
    let packagePath = path_1.default.join(parentNodeModulesPath, packageName);
    if (await util.fileExists(packagePath)) {
        return packagePath;
    }
    const pathComposite = [parentNodeModulesPath, "..", ".."];
    for (let i = 0; i < parentName.split("/").length - 1; i++) {
        pathComposite.push("..");
    }
    pathComposite.push(packageName);
    packagePath = path_1.default.join(...pathComposite);
    if (await util.fileExists(packagePath)) {
        return packagePath;
    }
    packagePath = path_1.default.join(rootNodeModulesPath, packageName);
    if (await util.fileExists(packagePath)) {
        return packagePath;
    }
    return null;
}
async function readPackages(parentName, dependencies, depth, parentNodeModulesPath, configuration, rootNodeModulesPath, packages) {
    if (depth > 0 && configuration.direct) {
        return;
    }
    const getPackagePromises = new Array();
    for (const dependency of Object.keys(dependencies)) {
        getPackagePromises.push(getPackage(parentName, dependency, parentNodeModulesPath, configuration, depth, rootNodeModulesPath, packages));
    }
    await Promise.all(getPackagePromises);
}
async function getPackage(parentName, dependency, parentNodeModulesPath, configuration, depth, rootNodeModulesPath, packages) {
    const packagePath = await getInstalledPath(parentName, dependency, parentNodeModulesPath, rootNodeModulesPath);
    if (packagePath === null) {
        console.error(chalk_1.default.red(`Package "${dependency}" was not found. Confirm that all modules are installed.`));
        return;
    }
    const file = await util.readPackageJson(path_1.default.join(packagePath, PACKAGE_JSON));
    if (!file) {
        console.error(chalk_1.default.red(`Package "${dependency}" is empty and cannot be analyzed.`));
        return;
    }
    const license = await (0, license_1.getLicense)(file, packagePath);
    const pack = {
        license: license.name,
        licenseFile: license.path,
        name: dependency,
        path: packagePath,
        repository: (0, repository_1.getRepository)(file.repository),
        version: file.version,
    };
    if (alreadyAnalyzed(packages, pack)) {
        return;
    }
    packages.push(pack);
    if (file.dependencies) {
        await readPackages(dependency, file.dependencies, depth + 1, path_1.default.join(packagePath, NODE_MODULES), configuration, rootNodeModulesPath, packages);
    }
}
