// license-compliance/lib/program.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.processArgs = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const commander_1 = tslib_1.__importDefault(require("commander"));
const enumerations_1 = require("./enumerations");
const license_1 = require("./license");
let program;
function processArgs() {
    program = new commander_1.default.Command();
    program
        .exitOverride(() => process.exit(1))
        .name("license-compliance")
        .description("Analyzes licenses of installed NPM packages, assisting with compliance.")
        .option("-p, --production", "Analyzes only production dependencies.")
        .option("-d, --development", "Analyzes only development dependencies.")
        .option("-t, --direct", "Analyzes only direct dependencies (depth = 1).")
        .option("-f, --format <format>", "Report format, csv, text, or json (default = text).", verifyFormat)
        .option("-r, --report <report>", "Report type, summary or detailed (default = summary).", verifyReport)
        .option("-q, --query <licenses>", "Semicolon separated list of licenses to query. Must conform to SPDX specifications.", verifyQuery)
        .option("-a, --allow <licenses>", "Semicolon separated list of allowed licenses. Must conform to SPDX specifications.", verifyAllow)
        .option("-e, --exclude <packages>", "Semicolon separated list of packages to be excluded from analysis. Regex expressions are supported.", verifyExclude)
        .parse(process.argv);
    verifyIncompatibleArguments();
    return program.opts();
}
exports.processArgs = processArgs;
function help(errorMessage) {
    console.error(chalk_1.default.red("Error:"), errorMessage);
    console.info(program.help());
}
function verifyAllow(value) {
    return value
        .split(";")
        .map((license) => license.trim())
        .filter((license) => !!license)
        .map((license) => {
        if (!(0, license_1.isLicenseValid)(license)) {
            help(`Invalid --allow option "${license}"`);
        }
        return license;
    });
}
function verifyQuery(value) {
    return value
        .split(";")
        .map((license) => license.trim())
        .filter((license) => !!license)
        .map((license) => {
        if (!(0, license_1.isLicenseValid)(license) && license !== "UNKNOWN") {
            help(`Invalid --query option "${license}"`);
        }
        return license;
    });
}
function verifyExclude(value) {
    return value
        .split(";")
        .map((exclude) => exclude.trim())
        .filter((exclude) => !!exclude)
        .map((exclude) => {
        if (exclude.startsWith("/") && exclude.endsWith("/")) {
            return RegExp(exclude.substring(1, exclude.length - 1));
        }
        return exclude;
    });
}
function verifyFormat(value) {
    if (!Object.keys(enumerations_1.Formatter).includes(value)) {
        help(`Invalid --format option "${value}"`);
    }
    return value;
}
function verifyReport(value) {
    if (!Object.keys(enumerations_1.Report).includes(value)) {
        help(`Invalid --report option "${value}"`);
    }
    return value;
}
function verifyIncompatibleArguments() {
    const options = program.opts();
    if (options.production && options.development) {
        help('Options "--production" and "--development" cannot be used together');
    }
    if (options.query && options.allow) {
        help("Options '--allow' and '--query' cannot be used together");
    }
}
