// license-compliance/lib/formatters/xunit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Xunit = void 0;
const tslib_1 = require("tslib");
const xmlbuilder_1 = tslib_1.__importDefault(require("xmlbuilder"));
class Xunit {
    constructor() {
        this.TEST_SUITES_NAME = "License Compliance";
        this.TEST_CASE_ERROR_TYPE = "License Compliance Error";
    }
    detail(packages) {
        const xmlContent = this.serializeObjectToXml({
            testsuites: this.formatDetailedTestSuites(packages),
        });
        console.info(xmlContent);
    }
    summary(licenses) {
        const xmlContent = this.serializeObjectToXml({
            testsuites: this.formatSummaryTestSuites(licenses),
        });
        console.info(xmlContent);
    }
    formatDetailedTestSuites(packages) {
        const detailedList = this.groupPackagesByLicense(packages);
        const totalNumberOfFailures = detailedList.reduce((sum, license) => sum + license.packages.length, 0);
        return {
            "@name": this.TEST_SUITES_NAME,
            "@tests": totalNumberOfFailures,
            "@errors": 0,
            "@failures": totalNumberOfFailures,
            testsuite: detailedList.map((license) => ({
                "@name": license.name,
                "@tests": license.packages.length,
                "@errors": 0,
                "@failures": license.packages.length,
                testcase: license.packages.map((packageInformation) => ({
                    "@name": `${packageInformation.name}@${packageInformation.version}`,
                    "@path": packageInformation.path,
                    failure: {
                        "@type": this.TEST_CASE_ERROR_TYPE,
                        "#text": `Package "${packageInformation.name}@${packageInformation.version}" uses non compliant license "${packageInformation.license}"`,
                    },
                })),
            })),
        };
    }
    formatSummaryTestSuites(licenses) {
        return {
            "@name": this.TEST_SUITES_NAME,
            "@tests": licenses.length,
            "@errors": 0,
            "@failures": licenses.length,
            testsuite: licenses.map((license) => ({
                "@name": license.name,
                "@tests": 1,
                "@errors": 0,
                "@failures": 1,
                testcase: [
                    {
                        "@name": license.name,
                        failure: {
                            "@type": this.TEST_CASE_ERROR_TYPE,
                            "#text": `${license.count} package${license.count > 1 ? "s" : ""} use non compliant license "${license.name}"`,
                        },
                    },
                ],
            })),
        };
    }
    groupPackagesByLicense(packages) {
        return packages.reduce((licenses, packageInformation) => {
            const licenseIndex = licenses.findIndex((license) => license.name === packageInformation.license);
            if (licenseIndex >= 0) {
                licenses[licenseIndex].packages.push(packageInformation);
            }
            else {
                licenses.push({
                    name: packageInformation.license,
                    packages: [packageInformation],
                });
            }
            return licenses;
        }, []);
    }
    serializeObjectToXml(object) {
        return xmlbuilder_1.default
            .create(object, {
            stringify: {
                attEscape: this.escapeString,
                textEscape: this.escapeString,
            },
        })
            .dec("1.0", "UTF-8")
            .end({
            pretty: true,
        });
    }
    escapeString(text) {
        return text
            .replace(/&/g, "&amp;")
            .replace(/"/g, "&quot;")
            .replace(/'/g, "&apos;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;");
    }
}
exports.Xunit = Xunit;
